lib/tests/tests.c

```diff
@@ -2084,6 +2084,122 @@ test_gene_conversion_simulation(void)
+    CU_ASSERT_FATAL(ret >= 0);
+    /* Site has two mutations */
+    ret = tsk_mutation_table_add_row(&tables.mutations, 0, 0, -1, "C", 1, NULL, 0);
+    CU_ASSERT_FATAL(ret >= 0);
+    ret = tsk_mutation_table_add_row(&tables.mutations, 0, 1, -1, "C", 1, NULL, 0);
+    CU_ASSERT_FATAL(ret >= 0);
+    ret = tsk_treeseq_init(&ts, &tables, TSK_BUILD_INDEXES);
+    CU_ASSERT_EQUAL_FATAL(ret, 0);
+
+
+    /* Ne <= 0 is an error */
+    ret = msp_log_likelihood_arg(&ts, 1, 0, &lik);
+    CU_ASSERT_EQUAL_FATAL(ret, MSP_ERR_BAD_POPULATION_SIZE);
+    ret = msp_log_likelihood_arg(&ts, -1, 0, &lik);
+    CU_ASSERT_EQUAL_FATAL(ret, MSP_ERR_BAD_POPULATION_SIZE);
+
+    ret = msp_unnormalised_log_likelihood_mut(&ts, 0, &lik);
+    CU_ASSERT_EQUAL_FATAL(ret, 0);
+    CU_ASSERT_EQUAL_FATAL(lik, -DBL_MAX);
+
+    ret = msp_unnormalised_log_likelihood_mut(&ts, 1, &lik);
+    CU_ASSERT_EQUAL_FATAL(ret, MSP_ERR_BAD_PARAM_VALUE);
+
+    tsk_treeseq_free(&ts);
+    tsk_table_collection_free(&tables);
+}
+
+static void
+test_likelihood_zero_edges(void)
+{
+    int ret;
+    double lik = 0;
+    tsk_table_collection_t tables;
+    tsk_treeseq_t ts;
+    ret = tsk_table_collection_init(&tables, 0);
+    CU_ASSERT_EQUAL_FATAL(ret, 0);
+    tables.sequence_length = 1;
+
+    ret = tsk_node_table_add_row(&tables.nodes, TSK_NODE_IS_SAMPLE, 0.0, TSK_NULL,
+            TSK_NULL, NULL, 0);
+    CU_ASSERT_FATAL(ret >= 0);
+    ret = tsk_node_table_add_row(&tables.nodes, TSK_NODE_IS_SAMPLE, 0.0, TSK_NULL,
+            TSK_NULL, NULL, 0);
+    CU_ASSERT_FATAL(ret >= 0);
+
+    ret = tsk_treeseq_init(&ts, &tables, TSK_BUILD_INDEXES);
+    CU_ASSERT_EQUAL_FATAL(ret, 0);
+
+    /* Zero edges should give a likelihood of zero */
+    ret = msp_log_likelihood_arg(&ts, 1, 1, &lik);
+    CU_ASSERT_EQUAL_FATAL(ret, 0);
+    CU_ASSERT_EQUAL_FATAL(lik, 0);
+
+    /* Zero mutations, so we get zero. */
+    ret = msp_unnormalised_log_likelihood_mut(&ts, 0, &lik);
+    CU_ASSERT_EQUAL_FATAL(ret, 0);
+    CU_ASSERT_EQUAL_FATAL(lik, 0);
+    ret = msp_unnormalised_log_likelihood_mut(&ts, 1, &lik);
+    CU_ASSERT_EQUAL_FATAL(ret, 0);
+    CU_ASSERT_EQUAL_FATAL(lik, 0);
+
+    tsk_treeseq_free(&ts);
+
+    /* Add in a site and a mutation */
+    ret = tsk_site_table_add_row(&tables.sites, 0.1, "A", 1, NULL, 0);
+    CU_ASSERT_FATAL(ret >= 0);
+    ret = tsk_mutation_table_add_row(&tables.mutations, 0, 0, -1, "C", 1, NULL, 0);
+    CU_ASSERT_FATAL(ret >= 0);
+    ret = tsk_treeseq_init(&ts, &tables, TSK_BUILD_INDEXES);
+    CU_ASSERT_EQUAL_FATAL(ret, 0);
+
+    /* Zero edges should give a likelihood of zero */
+    ret = msp_log_likelihood_arg(&ts, 1, 1, &lik);
+    CU_ASSERT_EQUAL_FATAL(ret, 0);
+    CU_ASSERT_EQUAL_FATAL(lik, 0);
+    ret = msp_log_likelihood_arg(&ts, 0, 1, &lik);
+    CU_ASSERT_EQUAL_FATAL(ret, 0);
+    CU_ASSERT_EQUAL_FATAL(lik, 0);
+
+    /* TODO why do we need this special case? Is it an error or does this
+     * return value have some significance? */
```

Comment:
This is returning a zero likelihood (-infinite log likelihood) when there are in the ARG but the putative mutation rate is zero. I was future-proofing for things like grid search or MCMC for parameter values.

Reply:
Ah, I see, thanks.. I'll update the code (we should document this when we're writing up the docs too).

Reply:
Should we also return -inf with a zero recombination rate? Seems like a reasonable thing to do?

Reply:
Yes, the documentation is on my to-do list, and finally getting rather high on it too.

Reply:
> Should we also return -inf with a zero recombination rate? Seems like a reasonable thing to do?

-inf when recombination rate is zero and there are break points in the ARG. If there are no break points, then the existing implementation will return a finite number when the recombination rate is zero.

Reply:
> I'll update the code (we should document this when we're writing up the docs too).

If you're updating the code, it would make sense for you to delete this comment while you're in there.